File: 2024/typescript/day03/day03.ts
const getMemory = async (filePath: string) => {
  const file = Bun.file(filePath);
  return await file.text();
};

const part2 = (memory: string) => {
  const reValid = /mul\((\d{1,3}),(\d{1,3})\)|don\'t\(\)|do\(\)/g;
  let match: RegExpExecArray | null;
  let enabled = true;
  let total = 0;
  while ((match = reValid.exec(memory))) {
    const val = match[0];
    if (val === "do()") enabled = true;
    else if (val === "don't()") enabled = false;
    else {
      if (enabled) {
        total += Number(match[1]) * Number(match[2]);
      }
    }
  }
  return total;
};

const part1 = (memory: string) => {
  const reValid = /mul\((\d{1,3}),(\d{1,3})\)/g;
  let match: RegExpExecArray | null;
  let total = 0;
  while ((match = reValid.exec(memory))) {
    total += Number(match[1]) * Number(match[2]);
  }
  return total;
};

const mem = await getMemory("day03/day03.input");
console.log("Part 1: ", part1(mem));
console.log("Part 2: ", part2(mem));

File: 2024/day03/day03.py
import re

with open("day03.input", "r") as f:
  memory = f.read()

def part2(mem: str):
  re_valid = r"mul\((\d{1,3}),(\d{1,3})\)|don\'t\(\)|do\(\)"
  enabled = True
  total = 0
  for match in re.finditer(re_valid, mem):
    val = match.group(0)
    if val == "do()":
      enabled = True
    elif val == "don't()":
      enabled = False
    else:
      if enabled:
        a, b = int(match.group(1)), int(match.group(2))
        total += (a * b)
  return total

def part1(mem: str):
  re_valid = r"mul\((\d{1,3}),(\d{1,3})\)"
  total = 0
  for match in re.finditer(re_valid, mem):
    a, b = int(match.group(1)), int(match.group(2))
    total += (a * b)
  return total

print('Part 1: {}'.format(part1(memory)))
print('Part 2: {}'.format(part2(memory)))